File: code/dobot_quad_sdk-main/low_level/python/e6_bms_state_sub.py
import dds_middleware_python as dds
import time
import sys


count = 0
last_time = time.time()


def lower_state_callback(state):
    global count, last_time
    count += 1

    now = time.time()
    if (now - last_time) * 1000 < 500:  # Only output every 500ms
        return
    last_time = now

    bms = state.bms_state()

    sys.stdout.write("\r")
    sys.stdout.write(f"Received BMS State #{count}\n")
    sys.stdout.write(f"Battery Level: {bms.battery_level()}\n")
    sys.stdout.write("\n")
    sys.stdout.flush()


def main():
    middleware = dds.PyDDSMiddleware("config/dds_config.yaml")
    middleware.subscribeLowerState("rt/lower/state", lower_state_callback)

    while True:
        time.sleep(1)


if __name__ == "__main__":
    main()

File: code/dobot_quad_sdk-main/low_level/cpp/e6_bms_state_sub.cc
#include <iostream>
#include <memory>
#include <thread>
#include <chrono>
#include "dds_middleware.hpp"
#include "lower_state.hpp"

using namespace dobotmh4::msg::dds_;

void lowerStateCallback(const LowerState_& state) {
    static int count = 0;
    static auto last_time = std::chrono::steady_clock::now();
    count++;
    
    auto now = std::chrono::steady_clock::now();
    if (std::chrono::duration_cast<std::chrono::milliseconds>(now - last_time).count() < 500) {
        return;  // Only output every 500ms
    }
    last_time = now;
    
    const BmsState_& bms = state.bms_state();
    
    std::cout << "\r\033[K";
    std::cout << "Received BMS State #" << count << std::endl
              << "Battery Level: " << bms.battery_level() << std::endl
              << "Battery ID: " << bms.bat_id() << std::endl
              << "BMS work time: " << bms.bms_work_time() << std::endl
              << "BMS current: " << bms.battery_now_current() << std::endl
              << std::endl
              << std::flush;
}

int main() {
    std::shared_ptr<dds_middleware::DDSMiddleware> middleware = 
        std::make_shared<dds_middleware::DDSMiddleware>(0);

    auto lower_state_sub = middleware->create_subscription<LowerState_>(
        "rt/lower/state", 
        lowerStateCallback,
        dds_middleware::QoSProfile::SensorData()
    );

    while (true) {
        std::this_thread::sleep_for(std::chrono::seconds(1));
    }
    return 0;
}